src/app/shared/entry/GA4GHFiles.state.service.ts

```diff
@@ -0,0 +1,123 @@
+/*
+ *     Copyright 2018 OICR
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License")
+ *     you may not use this file except in compliance with the License
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ */
+import { Injectable } from '@angular/core';
+import { BehaviorSubject, combineLatest, Observable } from 'rxjs';
+import { map } from 'rxjs/operators';
+
+import { DescriptorType } from '../enum/descriptorType.enum';
+import { GA4GHService, ToolFile } from '../swagger';
+
+/**
+ * This file serves as the state for the GA4GH files endpoint response
+ * Contains both actions and store
+ * @export
+ * @class Ga4ghFilesStateService
+ */
+@Injectable()
+export class GA4GHFilesStateService {
+  public toolFiles$: Observable<Array<ToolFile>>;
+
+  // TODO: Don't use BehaviorSubject, there should be a cleaner way with some Observable cleverness
+  public cwlToolFiles$ = new BehaviorSubject<Array<ToolFile>>(null);
+  public wdlToolFiles$ = new BehaviorSubject<Array<ToolFile>>(null);
+  public nflToolFiles$ = new BehaviorSubject<Array<ToolFile>>(null);
+  public containerToolFiles$: Observable<Array<ToolFile>>;
+  public descriptorToolFiles$: Observable<Array<ToolFile>>;
+  public testToolFiles$: Observable<Array<ToolFile>>;
+  constructor(private ga4ghService: GA4GHService) {
+    this.toolFiles$ = combineLatest(this.cwlToolFiles$, this.wdlToolFiles$, this.nflToolFiles$,
+      (cwlToolFiles, wdlToolFiles, nflToolFiles) => {
+        return this.mergeArray(cwlToolFiles, wdlToolFiles, nflToolFiles);
+      });
+    this.containerToolFiles$ = this.toolFiles$.pipe(map((toolFiles: Array<ToolFile>) => {
```

Comment:
these three look really repetitive, potential for refactoring?

Reply:
would be too difficult

Reply:
Add a TODO comment then for future cleanup and to warn future developers about evolving just one in a way that might make future refactoring more difficult. In Java you'd probably use a lambda, supplier, etc.
https://github.com/TypeStrong/atom-typescript/issues/848